### PR TITLE
Speedup refreshing

### LINKED ISSUE #141: Multiple calls to setDiff when triggering refresh
When triggering refresh manually 5 times DoubleTreeWidget::setDiff will be called.

### PR DESCRIPTION
**Pull request by Murmele**
delete line 1-3 from the drawio image. So less DoubleTreeWidgetd::setDiff() are called which speeds up everything

- fixes also the restoring of the selection in the treeview
- [x] Total test run is longer than on Master. Check!

fixes #141

### CODE REVIEW

**Comment by Murmele** [comment]
Not yet working:
merge without conflicts does not jump to the newly created commit

**Comment by Murmele** [comment]
> Not yet working:
> 
>     * [x]  merge without conflicts does not jump to the newly created commit
> 
>     * [ ]  When workdir is changed (New file added) Gittyup does not refresh correctly
> 
>     * [ ]  Also when removing untracked files (probably same as above)

@exactly-one-kas do you see also the second case. On Linux it refreshes as normal

**Comment by Murmele** [comment]
Validation script:

```python
import os
import re
import datetime

def parse(filename: str, master: bool) -> []:

    projects = []

    print(f"Parse: {filename}")
    if not os.path.isfile(filename):
        raise Exception(f"File not found: {filename}")

    pattern_Datetime = "QDateTime\((?P<time>[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}.[0-9]*) CET Qt::LocalTime\)"
    if master:
        arg = ""
    else:
        arg = "bool"

    pattern_openProject = 'static MainWindow\s*\*\s*MainWindow::open\(const QString\s*&\s*, bool\) ": " Open project:  "(?P<project_path>.*)"'
    pattern_Repoview_refresh = f'void RepoView::refresh\({arg}\) ": " time:  {pattern_Datetime}  Set diff counter:  (?P<counter>[0-9]*)'
    pattern_DoubleTreeWidget = f'virtual void DoubleTreeWidget::setDiff\(const git::Diff\s*&\s*, const QString\s*&\s*, const QString\s*&\s*\) ": " time:  {pattern_Datetime} Counter:  (?P<counter>[0-9]*)'
    pattern_DoubleTreeWidget_finished = f'virtual void DoubleTreeWidget::setDiff\(const git::Diff\s*&, const QString\s*&, const QString\s*&\s*\) ": " finished, time:  {pattern_Datetime} Counter:  (?P<counter>[0-9]*)'
    pattern_CommitListMousePress = f'virtual void CommitList::mousePressEvent\(QMouseEvent\s*\*\s*\) ": " time:  {pattern_Datetime}'
    pattern_Repoview_refresh = re.compile(pattern_Repoview_refresh)
    pattern_DoubleTreeWidget = re.compile(pattern_DoubleTreeWidget)
    pattern_CommitListMousePress = re.compile(pattern_CommitListMousePress)
    pattern_openProject = re.compile(pattern_openProject)
    pattern_DoubleTreeWidget_finished = re.compile(pattern_DoubleTreeWidget_finished)

    repoview_refresh_time = None
    refresh_initial_setDiffCounter = 0

    datetimeformat = "%Y-%m-%d %H:%M:%S.%f"

    matchCounter = 0
    with open(filename, "r") as f:
        lines = f.readlines()

        line_number = 0
        for line in lines:
            line_number += 1
            match = re.match(pattern_openProject, line)
            if match is not None:
                project_path = match["project_path"]
                projects.append({
                    "path": project_path,
                    "refreshes": [],
                })

            match = re.match(pattern_Repoview_refresh, line)
            if match is not None:
                #print("Refresh triggered")
                repoview_refresh_time = datetime.datetime.strptime(match["time"], datetimeformat)
                refresh_initial_setDiffCounter = int(match["counter"])
                t = {
                    "refresh_time": repoview_refresh_time,
                    "refresh_initial_setDiffCounter": refresh_initial_setDiffCounter,
                    "setdiffs": [],
                    "type": "repoview_refresh"
                }
                projects[-1]["refreshes"].append(t)
                continue


            match = re.match(pattern_CommitListMousePress, line)
            if match is not None:
                repoview_refresh_time = datetime.datetime.strptime(match["time"], datetimeformat)
                refresh_initial_setDiffCounter = 0
                t = {
                    "refresh_time": repoview_refresh_time,
                    "refresh_initial_setDiffCounter": refresh_initial_setDiffCounter,
                    "setdiffs": [],
                    "type": "commitlist_selection_changed"
                }
                projects[-1]["refreshes"].append(t)
                continue

            match = re.match(pattern_DoubleTreeWidget, line)
            if match is not None:
                time = datetime.datetime.strptime(match["time"], datetimeformat)
                counter = int(match["counter"])
                matchCounter += 1

                #print(f"Counter: {matchCounter}, Difference Time: {(time - repoview_refresh_time).total_seconds()*1000}ms. Count diff: {counter - refresh_initial_setDiffCounter}")

                assert(len(projects[-1]["refreshes"]) > 0)

                t = {
                    "totalCount": matchCounter,
                    "time_start": time,
                    "timediff_ms": (time - repoview_refresh_time).total_seconds()*1000,
                    "relativeCount": counter - refresh_initial_setDiffCounter,
                    "duration_setDiff": -1
                }
                projects[-1]["refreshes"][-1]["setdiffs"].append(t)
                continue

            match = re.match(pattern_DoubleTreeWidget_finished, line)
            if match is not None:
                time = datetime.datetime.strptime(match["time"], datetimeformat)
                counter = int(match["counter"])

                # print(f"Counter: {matchCounter}, Difference Time: {(time - repoview_refresh_time).total_seconds()*1000}ms. Count diff: {counter - refresh_initial_setDiffCounter}")

                assert (len(projects[-1]["refreshes"]) > 0)

                projects[-1]["refreshes"][-1]["setdiffs"][-1]["duration_setDiff"] = (time - projects[-1]["refreshes"][-1]["setdiffs"][-1]["time_start"]).total_seconds()*1000
                continue
    return projects

def hline() -> str:
    return "-----------------------------------\n"

def diffRow(diff: dict) -> str:
    return f'c:{diff["relativeCount"]}, t:{diff["timediff_ms"]}ms, d_setDiff:{diff["duration_setDiff"]}ms'

def rowString(columns: []):
    row = "|"
    for column in columns:
        row += f"{column:40}|"
    return row

def row(header: [], diffs: []) -> str:
    r = ""
    assert(len(header) == len(diffs))

    max_len = -1
    for d in diffs:
        length = len(d["setdiffs"])
        if length > max_len:
            max_len = length
    assert(max_len > 0)

    r += rowString(header) + "\n"
    for i in range(max_len - 1):
        row_contents = []
        for diff in diffs:
            row_content = ""
            if i < len(diff["setdiffs"]) - 1:
                row_content = diffRow(diff["setdiffs"][i])
            row_contents.append(row_content)

        r += rowString(row_contents) + "\n"

    row_contents = []
    for diff in diffs:
        row_content = diffRow(diff["setdiffs"][-1])
        row_contents.append(row_content)
    r += rowString(row_contents) + "\n"
    return r

def printTable(runs):
    """
    Create caller table
    :param runs:
    :return:
    """
    table = ""
    ref_logs = runs[0]["logs"]
    for p_index in range(len(ref_logs)):
        table += hline()
        table += hline()
        project_name = ref_logs[p_index]["path"]
        table += f"Open Project: {project_name}\n"

        for refreshes_index in range(len(ref_logs[p_index]["refreshes"])):
            table += hline()
            table += f"Refresh triggered: {ref_logs[p_index]['refreshes'][refreshes_index]['type']}\n"
            names = []
            values = []
            for run in runs:
                names.append(run["name"])
                values.append(run["logs"][p_index]["refreshes"][refreshes_index])
            table += row(names, values)
            table += hline()

    print(table)

if __name__ == "__main__":
    runs = [
        {
            "name": "OptimizedRefresh",
            "logs": parse("GittyupSpeedupRefresh.txt", False)
        },
        {
            "name": "Master",
            "logs": parse("GittyupMasterRefreshOutput.txt", True)
        }
    ]

    # check that all have the same number of open project
    num_proj = len(runs[0]["logs"])
    for run in runs:
        assert len(run["logs"]) == num_proj

    # print table for optimized refresh
    print(f"Run {runs[0]['name']}")
    printTable([runs[0]])

    # print table of current master
    print(f"Run {runs[1]['name']}")
    printTable([runs[1]])
```

File: src/ui/CommitList.h [hunk 1]
```diff
@@ -49,8 +49,11 @@ class CommitList : public QListView {
   void selectFirstCommit(bool spontaneous = false);
   bool selectRange(const QString &range, const QString &file = QString(),
                    bool spontaneous = false);
+  void suppressResetWalker(bool suppress);
+  bool isSuppressResetWalker();
```
Comment (exactly-one-kas) [inline src/ui/CommitList.h:53]: `isResetWalkerSuppressed` would be a better name

File: src/util/Debug.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#ifndef DEBUG_H
+#define DEBUG_H
+
+#include <QDebug>
+
+#define DEBUG 1
+#define DEBUG_REFRESH 1
```
Comment (exactly-one-kas) [inline src/util/Debug.h:7]: Wouldn't it be better to define these as parameters via CMAKE?

Comment (exactly-one-kas) [reply src/util/Debug.h:7]: Also: CMAKE defines `NDEBUG` for release builds, we could use that instead

File: test/rebase.cpp [hunk 3]
```diff
@@ -97,78 +97,80 @@ private slots:
...
-  auto *detailview = repoView->findChild<DetailView *>();
-  QVERIFY(detailview);
-  auto *abortRebaseButton = detailview->findChild<QPushButton *>("AbortRebase");
-  QVERIFY(abortRebaseButton);
-  auto *continueRebaseButton =
-      detailview->findChild<QPushButton *>("ContinueRebase");
-  QVERIFY(continueRebaseButton);
-  QCOMPARE(continueRebaseButton->isVisible(), false);
-  QCOMPARE(abortRebaseButton->isVisible(), false);
+  //  INIT_REPO("rebaseConflicts.zip", true);
```
Comment (exactly-one-kas) [inline test/rebase.cpp:100]: Why is this all disabled?

File: test/rebase.cpp [hunk 4]
```diff
@@ -599,243 +605,244 @@ void TestRebase::startRebaseContinueInCLIContinueGUI() {
...
-  // Check that buttons are visible
-  QCOMPARE(continueRebaseButton->isVisible(), false);
-  QCOMPARE(abortRebaseButton->isVisible(), false);
-
-  // Check call counters
-  QCOMPARE(rebaseFinished, 0);
-  QCOMPARE(rebaseAboutToRebase, 1);
-  QCOMPARE(rebaseCommitSuccess, 0);
-  QCOMPARE(rebaseConflict, 0);
+  //  INIT_REPO("rebaseConflicts.zip", true);
```
Comment (exactly-one-kas) [inline test/rebase.cpp:608]: This too

**Review by exactly-one-kas — CHANGES_REQUESTED** [review]

File: src/ui/CommitList.h (hunk 1 above)
Comment (Murmele) [reply src/ui/CommitList.h:53]: fixed in 8bdfca3f37c57a05356f9e1523d97ae80f16dc8f

File: src/util/Debug.h (hunk 2 above)
Comment (Murmele) [reply src/util/Debug.h]: first comment changed in ba02026813dca9709ff1482e12a7ab38e09d0e4e

File: test/rebase.cpp (hunk 3 above)
Comment (Murmele) [reply test/rebase.cpp]: Hmm I don't know. I enabled them again in 59846affcd58996b9061c13ebc8c9205614bc5cf. Thanks!

File: test/rebase.cpp (hunk 4 above)
Comment (Murmele) [reply test/rebase.cpp]: Hmm I don't know. I enabled them again in 59846affcd58996b9061c13ebc8c9205614bc5cf. Thanks!

File: src/util/Debug.h (hunk 2 above)
Comment (Murmele) [reply src/util/Debug.h]: Now in release mode no debug messages will be send out

**Pull request merged**